File: DSA/codechef/oct challenge/missing_num.cpp
#include<bits/stdc++.h>
using namespace std;

#define ll long long int 
ll th = 1e12;
ll find_num(ll base ,string x)
{
	ll s=0;
	ll c=0;
	ll rem=0;
	ll f=0;
	ll l=x.length();
	while(l>0)
	{
		if(x[l-1]>=65)
		{
			if(base>int(x[l-1]-55))
			{
				s+=pow(base,c)*int(x[l-1]-55);
				if( s>th )
					return -1;
			}
			else
			{
				return -1;
			}
		}
		else
		{
			if(base>int(x[l-1]-55))
			{
				s+=pow(base,c)*int(x[l-1]-48);
				if( s>th )
					return -1;
			}
			else
			{
				return -1;
			}
		}
		
		l--;
		c++;
		// cout<<"s"<<" "<<s<<endl;
	}
		return s;
}
ll find_base(ll final, string ans)
{
	string x=ans;
	ll f=0;
	ll z=0;
	
	for(ll j=2;j<=36;j++)
	{
		f=0;
		z=find_num(j,ans);
		if(z==final)
		{
			return j;
		}
		
	}
	
	
	return -1;
		
	

}
int  main()
{
	ll t;
	cin>>t;
	while(t--)
	{
		ll n;
		cin>>n;
		ll x;
		string y;
		ll a[n];
		string b[n];
		ll c1=0;
		ll c2=0;
		ll z;
		for(ll i=0;i<n;i++)
		{
			cin>>x;
			cin>>y;
			a[c1++]=x;
			b[c2++]=y;
		}
		ll base;
		for(ll i=0;i<n;i++)
		{
			if(a[i]!=-1)
			{
				base=a[i];
				x=find_num(base,b[i]);
				break;
			}

		}
		if(x==-1)
		{
			cout<<-1<<endl;
		}
		else
		{
			
			ll f=0;
				for(ll i=0;i<n;i++)
				{
					if(a[i]==-1)
					{
						z=find_base(x,b[i]);
						if(z!=-1)
						{
							a[i]=z;
						}
					}
				}
				for(ll i=0;i<n;i++)
				{
					if(a[i]==-1)
					{
						f=1;
					}
				}
				if(f!=1)
				{
					cout<<x<<endl;
				}
				else
				{
					cout<<"-1"<<endl;
				}
			
			//cout<<x<<endl;
		}
		

	}
}

File: DSA/codechef/oct challenge/missing_num.py

def find_num(base , x):
	l= len(x)-1
	sum = 0
	c=0
	while l>=0:
		v = 1
		if x[l]>='A':
			v = ord(x[l]) - 55
		else:
			v = ord(x[l]) - 48
		if v>=base:
			return -1
		sum = sum + v * (base**c)
		c += 1
		l -= 1
		if sum>10**12:
			return -1

	return sum




t= int(input())
while t>0:
	n =int(input())
	d = {}
	for z in range(n):
		b, s = input().split()
		b= int(b)
		d1 = {}
		if b==-1:

			for i in range(2,37):
				
				num=find_num(i,s)
				if num!=-1  and num not in d1:
					d1[num]=1
					if num not in d:
						d[num]=1
					else:
						d[num]+=1
		else:
			num=find_num(b,s)
			if num!=-1 and num not in d1:
				d1[num]=1
				if num not in d:
					d[num]=1
				else:
					d[num]+=1

	ans=-1
	for val in sorted(d.keys()):
		# print(val, d[val])
		if d[val]==n:
			ans=val
			break

	print(ans)
	t-=1


